ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaMain.cpp

```diff
@@ -98,4 +99,19 @@ string parameter_string(int n) {
   return string("");
 }
 
+string shellscript_evaluate(string command, size_t buffer_size) {
+  char buffer[buffer_size];
+  FILE *file = popen(command.c_str(), "r");
+  while (fgets(buffer, buffer_size, file) != NULL) {}
+  pclose(file);
+
+  if (buffer[strlen(buffer) - 1] == '\n')
+    buffer[strlen(buffer) - 1] = '\0';
+
+  char *result = (char *)malloc(strlen(buffer) + 1);
```

Comment:
This is a memory leak.

Reply:
You didn't tell me how to make it not leak memory. This could save me hours of Google searching.

Reply:
In general, don't use `malloc` (or even `new`) in C++. The number of times you need to call either function in a modern C++ application is essentially zero. And I didn't think I'd need to point out an alternative, because this buffer is completely unnecessary the way this code is implemented (this should have just read `return buffer;`) and unusable the way this code should be implemented: `strcpy` stops at the first null character, so your code, as written, *depends* on the string being null-terminated (which, fortunately, it is). But to support multi-line output of arbitrary size, you'll need to append multiple chunks to the output string, anyway.

Reply:
I just tried returning buffer which was how my code originally was before posted here and yeah
warning: address of local variable ‘buffer’ returned [-Wreturn-local-addr]

Reply:
oh i think i know why it's doing that with my code im testing it returns a char * not a std::string. silly me